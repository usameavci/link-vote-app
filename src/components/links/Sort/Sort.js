import find from "lodash/find";
import React, { useEffect, useState } from "react";
import { withRouter } from "react-router-dom";

import Url from "../../../utils/Url";
import CommonSelect from "../../common/Select/Select";

const Sort = ({ history, location }) => {
  const url = new Url(location);
  const { sort, order } = url.toObject({ only: ["sort", "order"] });

  const [options] = useState([
    {
      id: 1,
      label: "Creation Date",
      sort: "default",
    },
    {
      id: 2,
      label: "Most Voted (Z -> A)",
      sort: "vote",
      order: "desc",
    },
    {
      id: 3,
      label: "Less Voted (A -> Z)",
      sort: "vote",
      order: "asc",
    },
  ]);

  const defaultSelectedOption = find(options, { sort, order });
  const [selectedOption, setSelectedOption] = useState(defaultSelectedOption);

  useEffect(() => {
    setSelectedOption(find(options, { sort, order }));
  }, [options, sort, order]);

  const handleChange = (option) => {
    setSelectedOption(option);

    if (option.sort && option.order) {
      url.setParam("sort", option.sort);
      url.setParam("order", option.order);
    } else {
      url.deleteParam("sort");
      url.deleteParam("order");
    }

    history.push({ pathname: url.getPathName(), search: url.toString({ full: true }) });
  };

  return (
    <div data-testid="link-sort" className="link-sort">
      <CommonSelect value={selectedOption} options={options} onChange={handleChange} />
    </div>
  );
};

export default withRouter(Sort);
