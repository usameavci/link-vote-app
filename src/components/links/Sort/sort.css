.link-sort {
  min-width: 180px;
  position: relative;
  display: inline-block;
}

.link-sort.link-sort--opened .link-sort-label {
  border-radius: 4px 4px 0 0;
}

.link-sort-label {
  display: flex;
  cursor: pointer;
  border-radius: 4px;
  background-color: #f2f2f2;
  border: 1px solid #dbd6d6;
  justify-content: space-between;
}

.link-sort-label-inner,
.link-sort-label-icon {
  padding: 0.5rem;
}

.link-sort-label-icon {
  border-left: 1px solid #dbd6d6;
}

.link-sort-options {
  left: 0;
  top: 100%;
  z-index: 1;
  width: 100%;
  display: none;
  margin-top: -1px;
  position: absolute;
  border: 1px solid #a2a2a2;
}

.link-sort-options .link-sort-option {
  padding: 0.5rem;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background-color: #e8e8e8;
  border-bottom: 1px solid #a2a2a2;
}

.link-sort-options .link-sort-option:hover {
  cursor: pointer;
  background-color: #d8d8d8;
}

.link-sort-options .link-sort-option:last-child {
  border-bottom: 0;
}
