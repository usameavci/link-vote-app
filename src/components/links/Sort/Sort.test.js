import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

import Sort from "./Sort";

describe("Links sort component tests", () => {
  it("renders without crashing", () => {
    const component = (
      <Router>
        <Sort />
      </Router>
    );

    const { getByTestId } = render(component);

    expect(getByTestId("link-sort")).toBeInTheDocument();
  });

  it("matches snapshots", () => {
    const component = (
      <Router>
        <Sort />
      </Router>
    );

    const tree = renderer.create(component);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
