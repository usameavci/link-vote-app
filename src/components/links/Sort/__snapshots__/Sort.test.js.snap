// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links sort component tests matches snapshots 1`] = `
<div
  className="link-sort"
  data-testid="link-sort"
>
  <div
    className="common-select"
    data-testid="common-select"
  >
    <div
      className="common-select-label"
      onClick={[Function]}
    >
      <span
        className="common-select-label-inner"
      >
        Creation Date
      </span>
      <span
        className="common-select-label-icon"
      >
        <i
          className="fa fa-caret-down"
        />
      </span>
    </div>
    <div
      className="common-select-options"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="common-select-option"
        onClick={[Function]}
      >
        Creation Date
      </div>
      <div
        className="common-select-option"
        onClick={[Function]}
      >
        Most Voted (Z -&gt; A)
      </div>
      <div
        className="common-select-option"
        onClick={[Function]}
      >
        Less Voted (A -&gt; Z)
      </div>
    </div>
  </div>
</div>
`;
